pkg/apis/pipeline/v1beta1/task_validation.go

```diff
@@ -359,6 +362,29 @@ func validateStepArrayUsage(step Step, prefix string, vars sets.String) *apis.Fi
 }
 
 func validateVariables(steps []Step, prefix string, vars sets.String) (errs *apis.FieldError) {
+	// validate that the variable name format follows the rules
+	// - Must only contain alphanumeric characters, hyphens (-), underscores (_), and dots (.)
+	// - Must begin with a letter or an underscore (_)
+	re := regexp.MustCompile(variableNameFormat)
+	invalidNames := []string{}
+	// Converting to sorted list here rather than just looping map keys
+	// because we want the order of items in vars to be deterministic for purpose of unit testing
+	for _, name := range vars.List() {
+		if !re.MatchString(name) {
+			invalidNames = append(invalidNames, name)
+		}
+	}
+
+	if len(invalidNames) != 0 {
+		return &apis.FieldError{
+			Message: fmt.Sprintf("The format of following variable names is invalid. %s", invalidNames),
+			// Empty path is required to make the `ViaField`, … work
+			Paths:   []string{""},
+			Details: "Names: \nMust only contain alphanumeric characters, hyphens (-), underscores (_), and dots (.)\nMust begin with a letter or an underscore (_)",
+		}
+	}
+
+	// We've checked param name format. Now, we want to check if param names are referenced correctly in each step
```

Comment:
Adding validation here just checks the name format of parameters declared under `taskSpec`. 
```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  generateName: name-format-
spec:
  taskSpec:
    params:
      - name: foo
        default: "lower"
    results:
      - name: echo-output
        description: "successful echo"
    steps:
      - name: echo-params
        image: bash
        script: |
          set -e
          echo $(params.0foo) | tee $(results.echo-output.path)
```

- When you echo with a non-existing param name but the name follows the [rules](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-parameters) , the validation can detect it's referencing a non-existing param and returns an error when the yaml file is applied.
- But when you echo with a non-existing param name but the name does not follow the [rules](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-parameters), the validation cannot detect it's referencing a non-existing param, and weird behaviour will still happen mentioned in the issue #4792. see the above example in yaml


So there are two tasks to solve #4792:
1. the code that validates against format, which has been implemented by the current code chunk 
2. the current code chunk can just live here or be moved to the beginning of [ValidateVariableP](https://github.com/tektoncd/pipeline/blob/f0f9e32da6b4145ede27a371900e853a3530c1ee/pkg/substitution/substitution.go#L48-L63).
3. now all parameter names format is valid. (if invalid names are found, errors will be returned)
4. change [extractVariablesFromString](https://github.com/tektoncd/pipeline/blob/f0f9e32da6b4145ede27a371900e853a3530c1ee/pkg/substitution/substitution.go#L149-L165) to make sure it extract ***ALL*** chars in `<NAME>` part of`$(params.<NAME>)` no matter if it follows the rules.  (Currently, it only extracts names that follow the rules because of this [regex](https://github.com/tektoncd/pipeline/blob/68f2a66c9b2ffcc23ef262bac95d1eb3856021cf/pkg/substitution/substitution.go#L28))
5. And then just check if the extracted string exists in the set [ValidateVariableP](https://github.com/tektoncd/pipeline/blob/f0f9e32da6b4145ede27a371900e853a3530c1ee/pkg/substitution/substitution.go#L48-L63).

Reply:
step 2 is implemented in #4833 !

Reply:
step 2 is solved in this pr https://github.com/tektoncd/pipeline/pull/4857